amqp_job.go

```diff
@@ -163,29 +164,86 @@ func (j *amqpJob) createStateUpdateBody(ctx gocontext.Context, state string) map
 }
 
 func (j *amqpJob) sendStateUpdate(ctx gocontext.Context, event, state string) error {
+	err := j.stateUpdatePool.Process(&amqpStateUpdatePayload{
+		job:   j,
+		ctx:   ctx,
+		event: event,
+		state: state,
+		body:  j.createStateUpdateBody(ctx, state),
+	})
+
+	if err == nil {
+		return nil
+	}
+
+	return err.(error)
+}
+
+func (j *amqpJob) SetupContext(ctx gocontext.Context) gocontext.Context { return ctx }
+
+func (j *amqpJob) Name() string { return "amqp" }
+
+type amqpStateUpdatePayload struct {
+	job   *amqpJob
+	ctx   gocontext.Context
+	event string
+	state string
+	body  map[string]interface{}
+}
+
+type amqpStateUpdateWorker struct {
+	stateUpdateChan *amqp.Channel
+	ctx             gocontext.Context
+	cancel          gocontext.CancelFunc
+}
+
+func (w *amqpStateUpdateWorker) Process(payload interface{}) interface{} {
+	p := payload.(*amqpStateUpdatePayload)
+	ctx, cancel := gocontext.WithCancel(p.ctx)
+
+	w.ctx = ctx
+	w.cancel = cancel
+
+	return w.sendStateUpdate(p)
+}
+
+func (w *amqpStateUpdateWorker) BlockUntilReady() {
```

Comment:
I had to stop and think for a bit why this doesn't do anything. I assume it's because the sendStateUpdate operation is already blocking, so we don't need it, but it might be good to add a comment here (if that's correct).

Reply:
that is correct, afaik `BlockUntilReady` exists for if you want to perform some background or warm-up work before accepting the next job, and ensures that the worker does not receive any jobs until then. i'll add a comment to clarify! :)